File: LeetCode/C++/trees-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findIndex(vector<int>& vec, int val) {
        for(int idx = 0; idx < vec.size(); idx++) {
            if(vec[idx] == val)
                return idx;
        }
        return -1;  // val not found (not needed in our case)
    }
    
    TreeNode* buildTreeRecursive(vector<int>& inorder, int in_start, int in_end,
                                 vector<int>& postorder, int post_start, int post_end) {
        if (in_end < in_start || post_end < post_start) {   // no node left
            return nullptr;
        }
        
        // Last element in Postorder should be root
        int rootVal = postorder[post_end];
        TreeNode* root = new TreeNode(rootVal);
        
        if (in_start == in_end || post_start == post_end) {
            return root;    
        }
        
        int root_idx = findIndex(inorder, rootVal);
        int n_left = root_idx - in_start;   // #nodes in left branch
        int n_right = in_end - root_idx;    // #nodes in right branch
        
        root->left = buildTreeRecursive(inorder, in_start, root_idx-1, 
                                        postorder, post_start, post_end-n_right-1);
        root->right = buildTreeRecursive(inorder, root_idx+1, in_end, 
                                         postorder, post_start+n_left, post_end-1);
        
        return root;
        
    }
    
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return buildTreeRecursive(inorder, 0, inorder.size()-1, 
                                  postorder, 0, postorder.size()-1);
    }
};

File: LeetCode/Python/trees-construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        return self.buildTreeInPostorder(inorder, 0, len(inorder)-1,\
                                    postorder, 0, len(postorder)-1)
        
    def buildTreeInPostorder(self, inorder: List[int], in_start: int, in_end: int,\
                            postorder: List[int], post_start: int, post_end: int) -> TreeNode:
        if in_end < in_start or post_end < post_start:  # no node left
            return None
        
        # Last element in Postorder should be root
        rootVal = postorder[post_end]
        root = TreeNode(rootVal)
        
        if in_start is in_end or post_start is post_end:
            return root
        
        # Find root in Inorder
        root_idx = inorder.index(rootVal)
        n_left = root_idx - in_start;   # #nodes in left branch
        n_right = in_end - root_idx;  # #nodes in right branch
        
        root.left = self.buildTreeInPostorder(inorder, in_start, root_idx-1,\
                                       postorder, post_start, post_end-n_right-1)
        root.right = self.buildTreeInPostorder(inorder, root_idx+1, in_end, 
                                         postorder, post_start+n_left, post_end-1)
        
        return root
